src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs

```diff
@@ -35,57 +38,113 @@ internal class CodeActionsHandler : AbstractRequestHandler<LSP.CodeActionParams,
-                // The edits will be calculated when the code action is either previewed or 
-                // invoked.
+                // Filter out code actions with options since they'll show dialogs and we can't remote the UI and the options.
+                if (codeFix.Action is CodeActionWithOptions)
+                {
+                    continue;
+                }
+
+                // Temporarily filter out suppress and configure code actions, as we'll later combine them under a top-level
+                // code action.
+                if (codeFix.Action is AbstractConfigurationActionWithNestedActions)
+                {
+                    continue;
+                }
+
+                results.Add(GenerateVSCodeAction(request, codeFix.Action, CodeActionKind.QuickFix));
+            }
 
-                // It's safe for the client to pass back the range/filename in the command to run
-                // on the server because the client will always re-issue a get code actions request
-                // before invoking a preview or running the command on the server.
+            foreach (var (action, _) in codeRefactorings.SelectMany(codeRefactoring => codeRefactoring.CodeActions))
+            {
+                // Filter out code actions with options since they'll show dialogs and we can't remote the UI and the options.
+                if (action is CodeActionWithOptions)
+                {
+                    continue;
+                }
 
-                result.Add(
-                    new LSP.Command
-                    {
-                        CommandIdentifier = RunCodeActionCommandName,
-                        Title = codeAction.Title,
-                        Arguments = new object[]
-                        {
-                                new RunCodeActionParams
-                                {
-                                    CodeActionParams = request,
-                                    Title = codeAction.Title
-                                }
-                        }
-                    });
+                results.Add(GenerateVSCodeAction(request, action, CodeActionKind.Refactor));
             }
 
-            return result.ToArrayAndFree();
+            // Special case (also dealt with specially in local Roslyn): 
+            // If we have configure/suppress code actions, combine them under one top-level code action.
+            var configureSuppressActions = codeFixes.Where(a => a.Action is AbstractConfigurationActionWithNestedActions);
+            if (configureSuppressActions.Any())
+            {
+                results.Add(GenerateVSCodeAction(request, new CodeActionWithNestedActions(
+                    CodeFixesResources.Suppress_or_Configure_issues,
+                    configureSuppressActions.Select(a => a.Action).ToImmutableArray(), true), CodeActionKind.QuickFix));
+            }
```

Comment:
i need @mavasani for this sort of stuff.

Reply:
still applies.

Reply:
Opened #45726 to track moving this logic to a uniform location (as per our earlier discussion).